### PR TITLE
CDAP-15836. Config validations

### PR DESCRIPTION
**Pull request by aonischuk**
Validate:
- input schema (presence and type of source fields)
- date formats
- target and source fields

https://issues.cask.co/browse/CDAP-15836

### CODE REVIEW

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+      }
+    }
+  }
+
+  @Nullable
+  public Schema getSchema() {
+    try {
+      return Strings.isNullOrEmpty(schema) ? null : Schema.parseJson(schema);
+    } catch (IOException e) {
+      throw new IllegalStateException("Unable to parse output schema: " + schema, e);
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: This should be `IllegalArgumentException`

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+            failureCollector.addFailure(String.format("Target field '%s' is not present in output schema",
+                                                      targetField), null)
+              .withConfigProperty(PROPERTY_TARGET_FIELDS)
+              .withOutputSchemaField(targetField);
+          }
+        }
+      }
+    } catch (IllegalStateException e) {
+      failureCollector.addFailure("Output schema cannot be parsed.", null)
+        .withStacktrace(e.getStackTrace());
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/date/DateTransformConfig.java:159]: `.withConfigProperty`

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+  public void validate(FailureCollector failureCollector, Schema inputSchema) {
+    try {
+      Schema outputSchema = getSchema();
+      if (!containsMacro(PROPERTY_TARGET_FIELDS)) {
+        List<String> targetFields = this.getTargetFields();
+        for (String targetField : targetFields) {
+          if (outputSchema.getField(targetField) == null) {
+            failureCollector.addFailure(String.format("Target field '%s' is not present in output schema",
+                                                      targetField), null)
+              .withConfigProperty(PROPERTY_TARGET_FIELDS)
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: this should use `.withConfigElement`

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java (hunk 2 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/date/DateTransformConfig.java:159]: There is no config property. It's just schema json cannot be parsed.

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java (hunk 1 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: Done

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java (hunk 3 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: Done

**Comment by aonischuk** [comment]
@CuriousVini addressed your comments. Can you please merge/review?

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java (hunk 2 above)
Comment (CuriousVini) [reply src/main/java/io/cdap/plugin/date/DateTransformConfig.java:159]: we use "schema" as config property.

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+  @VisibleForTesting
+  public DateTransformConfig(String sourceFields, @Nullable String sourceFormat,
+                             String targetFields, @Nullable String targetFormat,
+                             @Nullable String secondsOrMilliseconds, String schema) {
+    this.sourceFields = sourceFields;
+    this.sourceFormat = sourceFormat;
+    this.targetFields = targetFields;
+    this.targetFormat = targetFormat;
+    this.secondsOrMilliseconds = (secondsOrMilliseconds == null) ? "Milliseconds" : secondsOrMilliseconds;
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: Make `Milliseconds` a constant.

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+      }
+    }
+  }
+
+  public void validateDateFormat(FailureCollector failureCollector, String value, String propertyName) {
+    if (!containsMacro(propertyName)) {
+      try {
+        new SimpleDateFormat(value);
+      } catch (IllegalArgumentException ex) {
+        failureCollector.addFailure(String.format("Field '%s' contains invalid date pattern '%s'",
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: error message should end with a period

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+          .withConfigProperty(PROPERTY_SOURCE_FIELDS)
+          .withInputSchemaField(field);
+      } else {
+        Schema inputFieldSchema = inputField.getSchema();
+
+        if (inputFieldSchema.isNullable()) {
+          inputFieldSchema = inputFieldSchema.getNonNullable();
+        }
+
+        if (inputFieldSchema.getType() != Schema.Type.STRING && inputFieldSchema.getType() != Schema.Type.LONG) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: should check for logical types as well.

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+  private DateTransformConfig(Builder builder) {
+    this.sourceFields = builder.sourceFields;
+    this.sourceFormat = builder.sourceFormat;
+    this.targetFields = builder.targetFields;
+    this.targetFormat = builder.targetFormat;
+    this.secondsOrMilliseconds = builder.secondsOrMilliseconds;
+    this.schema = builder.schema;
+  }
+
+  public String getSourceFormat() {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: should add @Nullable

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    return sourceFormat;
+  }
+
+  public String getTargetFormat() {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: should add @Nullable

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+    return targetFormat;
+  }
+
+  public boolean isInSeconds() {
+    if (secondsOrMilliseconds == null) {
+      return false;
+    }
+    return secondsOrMilliseconds.equals("Seconds");
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: use a string constant

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+  }
+
+  public void validate(FailureCollector failureCollector, Schema inputSchema) {
+    try {
+      Schema outputSchema = getSchema();
+      if (!containsMacro(PROPERTY_TARGET_FIELDS)) {
+        List<String> targetFields = this.getTargetFields();
+        for (String targetField : targetFields) {
+          if (outputSchema.getField(targetField) == null) {
+            failureCollector.addFailure(String.format("Target field '%s' is not present in output schema",
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: period at the end of error message

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+          .withInputSchemaField(field);
+      } else {
+        Schema inputFieldSchema = inputField.getSchema();
+
+        if (inputFieldSchema.isNullable()) {
+          inputFieldSchema = inputFieldSchema.getNonNullable();
+        }
+
+        if (inputFieldSchema.getType() != Schema.Type.STRING && inputFieldSchema.getType() != Schema.Type.LONG) {
+          failureCollector.addFailure(String.format("Source field '%s' must be of type 'string' or 'long'" +
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: Error message: `Source Field '%s' is unexpected type '%s'.`
corrective action: `Supported types are 'string' or 'long'.`

**Review by CuriousVini — COMMENTED** [review]
A few comments

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java (hunk 4 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: Done

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java (hunk 5 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: Done

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java (hunk 6 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: Done

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java (hunk 9 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: Done

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java (hunk 10 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: Done

File: src/main/java/io/cdap/plugin/date/DateTransformConfig.java (hunk 11 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/date/DateTransformConfig.java]: Done

**Comment by aonischuk** [comment]
@CuriousVini thanks for your review. Can you please merge if LG? (ps. Addressed all the comments)

**Review by CuriousVini — APPROVED** [review]

**Pull request merged**